File: 04. LinkedList/06. Swap nodes in pairs/main.cpp
#include <bits/stdc++.h>
using namespace std;

struct ListNode {
    int val;
    ListNode* next;
    ListNode(int x) : val(x), next(NULL) {}
};

ListNode* swapPairs(ListNode* head) {
    if (!head || !head->next) return head;

    ListNode* newHead = head->next;
    ListNode* prev = NULL;

    while (head && head->next) {
        ListNode* first = head;
        ListNode* second = head->next;

        first->next = second->next;
        second->next = first;

        if (prev) prev->next = second;
        prev = first;
        head = first->next;
    }

    return newHead;
}

// Utility function to print a linked list
void printList(ListNode* head) {
    while (head) {
        cout << head->val << " -> ";
        head = head->next;
    }
    cout << "NULL\n";
}

int main() {
    ListNode* head = new ListNode(1);
    head->next = new ListNode(2);
    head->next->next = new ListNode(3);
    head->next->next->next = new ListNode(4);

    cout << "Original List: ";
    printList(head);

    head = swapPairs(head);

    cout << "List after swapping pairs: ";
    printList(head);

    return 0;
}

File: 04. LinkedList/06. Swap nodes in pairs/main.py
class ListNode:
    def __init__(self, val):
        self.val = val
        self.next = None

def swapPairs(head):
    if not head or not head.next:
        return head

    newHead = head.next
    prev = None

    while head and head.next:
        first = head
        second = head.next

        first.next = second.next
        second.next = first

        if prev:
            prev.next = second
        prev = first
        head = first.next

    return newHead

# Utility function to print a linked list
def printList(head):
    while head:
        print(head.val, end=" -> ")
        head = head.next
    print("NULL")

# Main function
def main():
    head = ListNode(1)
    head.next = ListNode(2)
    head.next.next = ListNode(3)
    head.next.next.next = ListNode(4)

    print("Original List:")
    printList(head)

    head = swapPairs(head)

    print("List after swapping pairs:")
    printList(head)

# Run main function
if __name__ == "__main__":
    main()
